### PR TITLE
[TS migration] Migrate `MoneyTemporaryForRefactorRequestConfirmationList.js` component to TypeScript

### PR DESCRIPTION
**Pull request by allgandalf**
### Details

This is a typescript migration issue, the file was migrated from `js` to `tsx`
### Fixed Issues

$ https://github.com/Expensify/App/issues/36130
PROPOSAL: https://github.com/Expensify/App/issues/36130#issuecomment-1943710520

### Tests

Same as QA step
- [x] Verify that no errors appear in the JS console

### Offline tests

Same as QA step
### QA Steps

1. Go to workspace chat/ 1:1 chat (Money Request)
2. Click on `+` icon
3. select Money Request
4. Perform all the steps and get on the final confirmation step
(This is a migration PR, so test out all options available on the final confirmation page)

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/110545952/79a50b37-87df-4cf9-8592-8684ea1392a3

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/110545952/5c661f5b-56ba-4333-a6bd-7924d5213dca

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/110545952/2406bdf1-d768-46bf-894f-2709a798bb93

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/110545952/cd16f4ea-9574-4949-85b7-9a81eefa3a65

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/110545952/708e0e04-e6ad-496e-a03b-9a01d3a0e621

</details>

<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/110545952/f2b2a8bc-39eb-4c46-806a-b699530dead3

</details>

### CODE REVIEW

**Comment by blazejkustra** [comment]
Is this ready for review? @GandalfGwaihir

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx [hunk 1]
```diff
@@ -465,15 +426,23 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
...
         }
-        return [...selectedParticipants, OptionsListUtils.getIOUConfirmationOptionsFromPayeePersonalDetail(personalDetailsOfPayee)];
-    }, [selectedParticipants, hasMultipleParticipants, personalDetailsOfPayee]);
+
+        const myIOUAmount = IOUUtils.calculateAmount(selectedParticipants.length, iouAmount, iouCurrencyCode ?? '', true);
+        return [
+            ...selectedParticipants,
+            OptionsListUtils.getIOUConfirmationOptionsFromPayeePersonalDetail(
+                personalDetailsOfPayee,
+                iouAmount > 0 ? CurrencyUtils.convertToDisplayString(myIOUAmount, iouCurrencyCode) : '',
```
Comment (blazejkustra) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:442]: Why was a second argument added here? This changes the logic quite a bit

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx [hunk 2]
```diff
@@ -620,13 +587,14 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
...
             <>
-                {!_.isEmpty(formError) && (
+                {formError && (
                     <FormHelpMessage
                         style={[styles.ph1, styles.mb2]}
                         isError
                         message={formError}
                     />
                 )}
+
```
Comment (blazejkustra) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:597]:
```suggestion
```

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx [hunk 3]
```diff
@@ -620,13 +587,14 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
 
         return (
             <>
-                {!_.isEmpty(formError) && (
+                {formError && (
```
Comment (blazejkustra) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:590]: LEt's cast to boolean here:
```suggestion
                {!!formError && (
```

[context](https://dev.to/maafaishal/avoid-operator-for-conditional-rendering-in-react-2de)

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx [hunk 4]
```diff
@@ -774,17 +762,24 @@ function MoneyTemporaryForRefactorRequestConfirmationList({
...
             isSupplementary: !isCategoryRequired,
         },
-        ..._.map(policyTagLists, ({name}, index) => ({
+        ...policyTagLists.map(({name}, index) => ({
             item: (
                 <MenuItemWithTopDescription
                     key={name}
                     shouldShowRightIcon={!isReadOnly}
-                    title={TransactionUtils.getTagForDisplay(transaction, index)}
+                    title={TransactionUtils.getTagForDisplay(transaction ?? null, index)}
```
Comment (blazejkustra) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:770]: I think we don't have to default to null for transactions, wdyt? Same for other places in this file
```suggestion
                    title={TransactionUtils.getTagForDisplay(transaction, index)}
```

File: src/components/PopoverMenu.tsx [hunk 5]
```diff
@@ -18,7 +18,7 @@ import Text from './Text';
 
 type PopoverMenuItem = {
     /** An icon element displayed on the left side */
-    icon: IconAsset;
+    icon?: IconAsset;
```
Comment (blazejkustra) [inline src/components/PopoverMenu.tsx:21]: Do we need to make icons optional in this file and in others?

**Review by blazejkustra — CHANGES_REQUESTED** [review]
The translate error happens because params for each translation key is different:
iou.splitAmount:
```ts
type RequestAmountParams = {amount: string};
```

iou.requestAmount
```ts
type SplitAmountParams = {amount: number};
```

I think you can adjust one so they use the same type

**Comment by blazejkustra** [comment]
@GandalfGwaihir Bumping this 😄

**Comment by blazejkustra** [comment]
But these are different components, right? I think we can proceed with this migration as usual

**Review by blazejkustra — APPROVED** [review]
Code looks good to me! 😄

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx [hunk 6]
```diff
@@ -25,241 +25,200 @@ import * as ReceiptUtils from '@libs/ReceiptUtils';
...
+    reportActionID?: string;
 };
 
 function MoneyTemporaryForRefactorRequestConfirmationList({
-    bankAccountRoute,
-    canModifyParticipants,
-    currentUserPersonalDetails,
-    hasMultipleParticipants,
-    hasSmartScanFailed,
+    transaction = null,
```
[the remark this answers is not in the document]
Comment (blazejkustra) [reply src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:164]: It's not a blocker for me, so no need to take an action 👍

**Comment by fabioh8010** [comment]
Hey @hoangzinh Is your review done here? If yes, could you complete the Reviewer Checklist? Thanks

**Comment by blazejkustra** [comment]
@GandalfGwaihir Did you find the cause of this bug?

**Comment by fabioh8010** [comment]
@GandalfGwaihir any news here?

**Review by hoangzinh — COMMENTED** [review]
Looks good. Just one feedback

**Review by hoangzinh — APPROVED** [review]
LGTM

**Review by hoangzinh — APPROVED** [review]

**Comment by fabioh8010** [comment]
@GandalfGwaihir @hoangzinh Looks like TS checks is passing on `main` now, so you can just update the branch and it should be enough!

**Comment by fabioh8010** [comment]
@GandalfGwaihir I've checked your PR and here is my suggestion to fix the TS issues:

```diff
diff --git a/src/libs/ReportUtils.ts b/src/libs/ReportUtils.ts
index 85f5c414db..ad4632f96f 100644
--- a/src/libs/ReportUtils.ts
+++ b/src/libs/ReportUtils.ts
@@ -418,7 +418,7 @@ type OptionData = {
     notificationPreference?: NotificationPreference | null;
     isDisabled?: boolean | null;
     name?: string | null;
-    isSelfDM?: boolean | null;
+    isSelfDM?: boolean;
     reportID?: string;
     enabled?: boolean;
     data?: Partial<TaxRate>;
diff --git a/src/pages/iou/request/step/IOURequestStepConfirmation.tsx b/src/pages/iou/request/step/IOURequestStepConfirmation.tsx
index 897af48f05..da12e30a00 100644
--- a/src/pages/iou/request/step/IOURequestStepConfirmation.tsx
+++ b/src/pages/iou/request/step/IOURequestStepConfirmation.tsx
@@ -2,7 +2,6 @@ import React, {useCallback, useEffect, useMemo, useRef, useState} from 'react';
 import {View} from 'react-native';
 import type {OnyxEntry} from 'react-native-onyx';
 import {withOnyx} from 'react-native-onyx';
-import type {ValueOf} from 'type-fest';
 import FullScreenLoadingIndicator from '@components/FullscreenLoadingIndicator';
 import HeaderWithBackButton from '@components/HeaderWithBackButton';
 import * as Expensicons from '@components/Icon/Expensicons';
@@ -30,11 +29,12 @@ import ROUTES from '@src/ROUTES';
 import type SCREENS from '@src/SCREENS';
 import type {Policy, PolicyCategories, PolicyTagList} from '@src/types/onyx';
 import type {Participant} from '@src/types/onyx/IOU';
+import type {PaymentMethodType} from '@src/types/onyx/OriginalMessage';
 import type {Receipt} from '@src/types/onyx/Transaction';
 import type {WithFullTransactionOrNotFoundProps} from './withFullTransactionOrNotFound';
 import withFullTransactionOrNotFound from './withFullTransactionOrNotFound';
-import withWritableReportOrNotFound from './withWritableReportOrNotFound';
 import type {WithWritableReportOrNotFoundProps} from './withWritableReportOrNotFound';
+import withWritableReportOrNotFound from './withWritableReportOrNotFound';
 
 type IOURequestStepConfirmationOnyxProps = {
     /** The policy of the report */
@@ -97,7 +97,7 @@ function IOURequestStepConfirmation({
             transaction?.participants?.map((participant) => {
                 const participantAccountID = participant.accountID ?? 0;
                 return participantAccountID ? OptionsListUtils.getParticipantsOption(participant, personalDetails) : OptionsListUtils.getReportOption(participant);
-            }),
+            }) ?? [],
         [transaction?.participants, personalDetails],
     );
     const isPolicyExpenseChat = useMemo(() => ReportUtils.isPolicyExpenseChat(ReportUtils.getRootParentReport(report)), [report]);
@@ -422,11 +422,9 @@ function IOURequestStepConfirmation({
 
     /**
      * Checks if user has a GOLD wallet then creates a paid IOU report on the fly
-     *
-     * @param {String} paymentMethodType
      */
     const sendMoney = useCallback(
-        (paymentMethodType: ValueOf<typeof CONST.IOU.PAYMENT_TYPE>) => {
+        (paymentMethod: PaymentMethodType | undefined) => {
             const currency = transaction?.currency;
 
             const trimmedComment = transaction?.comment?.comment ? transaction.comment.comment.trim() : '';
@@ -437,12 +435,12 @@ function IOURequestStepConfirmation({
                 return;
             }
 
-            if (paymentMethodType === CONST.IOU.PAYMENT_TYPE.ELSEWHERE) {
+            if (paymentMethod === CONST.IOU.PAYMENT_TYPE.ELSEWHERE) {
                 IOU.sendMoneyElsewhere(report, transaction.amount, currency, trimmedComment, currentUserPersonalDetails.accountID, participant);
                 return;
             }
 
-            if (paymentMethodType === CONST.IOU.PAYMENT_TYPE.EXPENSIFY) {
+            if (paymentMethod === CONST.IOU.PAYMENT_TYPE.EXPENSIFY) {
                 IOU.sendMoneyWithWallet(report, transaction.amount, currency, trimmedComment, currentUserPersonalDetails.accountID, participant);
             }
         },
@@ -490,12 +488,11 @@ function IOURequestStepConfirmation({
                     />
                     {isLoading && <FullScreenLoadingIndicator />}
                     <View style={[styles.flex1, isLoading && styles.opacity0]}>
-                        {/* @ts-expect-error TODO: Remove this once MoneyRequestConfirmationList (https://github.com/Expensify/App/issues/36130) is migrated to TypeScript. */}
                         <MoneyRequestConfirmationList
                             transaction={transaction}
                             hasMultipleParticipants={iouType === CONST.IOU.TYPE.SPLIT}
                             selectedParticipants={participants}
-                            iouAmount={transaction?.amount}
+                            iouAmount={transaction?.amount ?? 0}
                             iouComment={transaction?.comment.comment ?? ''}
                             iouCurrencyCode={transaction?.currency}
                             iouIsBillable={transaction?.billable}

```

File: src/pages/iou/request/step/IOURequestStepConfirmation.tsx [hunk 7]
```diff
@@ -490,12 +488,11 @@ function IOURequestStepConfirmation({
                     />
                     {isLoading && <FullScreenLoadingIndicator />}
                     <View style={[styles.flex1, isLoading && styles.opacity0]}>
-                        {/* @ts-expect-error TODO: Remove this once MoneyRequestConfirmationList (https://github.com/Expensify/App/issues/36130) is migrated to TypeScript. */}
                         <MoneyRequestConfirmationList
                             transaction={transaction}
                             hasMultipleParticipants={iouType === CONST.IOU.TYPE.SPLIT}
                             selectedParticipants={participants}
-                            iouAmount={transaction?.amount}
+                            iouAmount={transaction?.amount ?? 0}
```
Comment (youssef-lr) [inline src/pages/iou/request/step/IOURequestStepConfirmation.tsx:495]: I don't think this is necessary, unless it was causing a TS error?

```suggestion
                            iouAmount={transaction?.amount}
```

Comment (blazejkustra) [reply src/pages/iou/request/step/IOURequestStepConfirmation.tsx:495]: It is causing an TS error @youssef-lr, I checked and it is a required prop

Comment (youssef-lr) [reply src/pages/iou/request/step/IOURequestStepConfirmation.tsx:495]: okay cool! Testing now locally, then I'll merge if nothing comes up

**Review by youssef-lr — APPROVED** [review]
LGTM!

**Comment by aldo-expensify** [comment]
I think this PR may be causing the deploy blocker: https://github.com/Expensify/App/issues/40476

**Comment by aldo-expensify** [comment]
Yeah, sorry I just noticed that too. Thanks for all the info!

**Pull request merged**